src/script/common/c_packer.cpp

```diff
@@ -0,0 +1,566 @@
+#include "log.h"
+#include "debug.h"
+#include "threading/mutex_auto_lock.h"
+
+extern "C" {
+#include <lauxlib.h>
+}
+
+/**** Helpers ****/
+
+// convert negative index to absolute position on Lua stack
+#define absidx(_L, _idx) (assert((_idx) < 0), lua_gettop(_L) + (_idx) + 1)
+
+// does the type put anything into PackedInstr::sdata?
+static inline bool uses_sdata(int type)
+{
+	switch (type) {
+		case LUA_TSTRING:
+		case LUA_TFUNCTION:
+		case LUA_TUSERDATA:
+			return true;
+		default:
+			return false;
+	}
+}
+
+// does the type put anything into PackedInstr::<union>?
+static inline bool uses_union(int type)
+{
+	switch (type) {
+		case LUA_TNIL:
+		case LUA_TSTRING:
+		case LUA_TFUNCTION:
+			return false;
+		default:
+			return true;
+	}
+}
+
+static inline bool can_set_into(int ktype, int vtype)
+{
+	switch (ktype) {
+		case LUA_TNUMBER:
+			return !uses_union(vtype);
+		case LUA_TSTRING:
+			return !uses_sdata(vtype);
+		default:
+			return false;
+	}
+}
+
+// is the key suitable for use with set_into?
+static inline bool suitable_key(lua_State *L, int idx)
+{
+	if (lua_type(L, idx) == LUA_TSTRING) {
+		// strings may not have a NULL byte (-> lua_setfield)
+		size_t len;
+		const char *str = lua_tolstring(L, idx, &len);
+		return strlen(str) == len;
+	} else {
+		assert(lua_type(L, idx) == LUA_TNUMBER);
+		// numbers must fit into an s32 and be integers (-> lua_rawseti)
+		lua_Number n = lua_tonumber(L, idx);
+		return std::floor(n) == n && n >= S32_MIN && n <= S32_MAX;
+	}
+}
+
+namespace {
+	// checks if you left any values on the stack, for debugging
+	class StackChecker {
+		lua_State *L;
+		int top;
+	public:
+		StackChecker(lua_State *L) : L(L), top(lua_gettop(L)) {}
+		~StackChecker() {
+			assert(lua_gettop(L) >= top);
+			if (lua_gettop(L) > top) {
+				rawstream << "Lua stack not cleaned up: "
+					<< lua_gettop(L) << " != " << top
+					<< " (false-positive if exception thrown)" << std::endl;
+			}
+		}
+	};
+
+	// Since an std::vector may reallocate this is the only safe way to keep
+	// a reference to a particular element.
+	template <typename T>
+	class VectorRef {
+		std::vector<T> *vec;
+		size_t idx;
+		VectorRef(std::vector<T> *vec, size_t idx) : vec(vec), idx(idx) {}
+	public:
+		static VectorRef<T> front(std::vector<T> &vec) {
+			return VectorRef(&vec, 0);
+		}
+		static VectorRef<T> back(std::vector<T> &vec) {
+			return VectorRef(&vec, vec.size() - 1);
+		}
+		T &operator*() { return (*vec)[idx]; }
+		T *operator->() { return &(*vec)[idx]; }
+	};
+
+	struct Packer {
+		PackInFunc fin;
+		PackOutFunc fout;
+	};
+
+	typedef std::pair<std::string, Packer> PackerTuple;
+}
+
+static inline auto emplace(PackedValue &pv, s16 type)
+{
+	pv.i.emplace_back();
+	auto ref = VectorRef<PackedInstr>::back(pv.i);
+	ref->type = type;
+	if (type == LUA_TTABLE) {
+		ref->uidata1 = 0;
+		ref->uidata2 = 0;
+	} else if (type == LUA_TUSERDATA) {
+		ref->ptrdata = nullptr;
+	} else if (type == INSTR_POP) {
+		ref->sidata2 = 0;
+	}
```

Comment:
It might make sense to use a switch/case here too, and add an error message in case the type is not handled.